ext/vulnsrc/redhat/ovalv2.go

```diff
@@ -0,0 +1,814 @@
+	for i := 1; i < len(affectedCpeList.Cpe); i++ {
+		cpeNames = append(cpeNames, affectedCpeList.Cpe[i])
+	}
+	return cpeNames, nil
+}
+
+// parse affected_cpe_list (first entry from CPE list should not be used because it doesn't come from Advisory configuration)
+func ParseCpeStructFromAffectedCpeList(affectedCpeList OvalV2Cpe) ([]CpeName, error) {
+	cpeStructs := []CpeName{}
+	if affectedCpeList.Cpe == nil || len(affectedCpeList.Cpe) < 2 {
+		return cpeStructs, errors.New("unparseable affected cpe list")
+	}
+	// parse and return any entries after the first cpe entry from the list
+	for i := 1; i < len(affectedCpeList.Cpe); i++ {
+		cpeStructs = append(cpeStructs, ParseCpeName(affectedCpeList.Cpe[i]))
+	}
+	return cpeStructs, nil
+}
+
+// parse cpe string
+func ParseCpeName(cpeNameString string) CpeName {
+	// cpe:/ {part} : {vendor} : {product} : {version} : {update} : {edition} : {language}
+	// remove the "cpe:/" prelude and split the cpe name string into its components
+	components := strings.Split(strings.Replace(cpeNameString, "cpe:/", "", 1), ":")
+	// components slice must contain 7 elements; append empty string to any missing elements
+	for i := len(components); i < 7; i++ {
+		components = append(components, "")
+	}
+	return CpeName{
+		Part:      components[0],
+		Vendor:    components[1],
+		Product:   components[2],
+		Version:   components[3],
+		Update:    components[4],
+		Edition:   components[5],
+		Language:  components[6],
+	}
+}
+
+// get advisories from the given oval document which were issued since the last update (based on db value)
+func ProcessAdvisoriesSinceLastDbUpdate(ovalDoc OvalV2Document, datastore database.Datastore) ([]ParsedAdvisory, error) {
+	sinceDate := DbLookupLastAdvisoryDate(datastore)
+	var advisories []ParsedAdvisory
+	for _, definition := range ovalDoc.DefinitionSet.Definitions {
+		// check if this entry has already been processed (based on its sha256 hash)
+		if IsAdvisorySinceDate(sinceDate, definition.Metadata.Advisory.Issued.Date) {
+			// this advisory was issued since the last advisory date in the database; add it
+			advisories = append(advisories, ParseAdvisory(definition, ovalDoc))
+		} else if IsAdvisorySameDate(sinceDate, definition.Metadata.Advisory.Issued.Date) {
+			parsedAdvisory := ParseAdvisory(definition, ovalDoc)
+			// advisory date is coarse (YYYY-MM-dd format) date only,
+			// so it's possible that we'll see an advisory multiple times within the same day;
+			// check the db in this case to be sure
+			if (!DbLookupIsAdvisoryProcessed(parsedAdvisory, datastore)) {
+				// this advisory id/version hasn't been processed yet; add it
+				advisories = append(advisories, parsedAdvisory)
+			}
+		}
+	}
+	// debug-only info
+	out, _ := xml.MarshalIndent(ovalDoc, " ", "  ")
+	log.Debug(string(out))
+
+	// update the db ky/value entry for the last advisory processed date (now, as coarse YYYY-MM-dd format)
+	DbStoreLastAdvisoryDate(time.Now().Format(AdvisoryDateFormat), datastore)
```

Comment:
The same applies here. When you store a new date to DB before parser successfully returns vulnerabilities it can happen that parses fails and in next iteration parses skips all advisories because the date doesn't match with reality.

Reply:
In `vulnsrc.UpdateResponse` there is a flag that allows you to pass some variables that are stored in the key-value table only when vulnerabilities are parsed successfully. Please use this variable. If you look at the rh-internal branch there is a change that allows returning more than one flag.